change default InMemoryTable sorting behavior

lizozom opened the pull request:
### Summary

This PR changes the default InMemoryTable sorting logic.
It now uses a three way sort on a field ASC -> DESC -> neutral

![ezgif-2-028937cd0fbb](https://user-images.githubusercontent.com/3016806/53348013-f0045d80-3922-11e9-9305-9f2dc0c589e1.gif)

Blocker for [EUIfication](https://github.com/elastic/kibana/issues/31599) of visualizations \ dashboards screens on Kibana

### Checklist

- [x] This was checked in mobile
- [ ] This was checked in IE11
- [x] This was checked in dark mode
- [x] Any props added have proper autodocs
- [x] Documentation examples were added
- [ ] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately
- [ ] This was checked for breaking changes and labeled appropriately
- [ ] Jest tests were updated or added to match the most common scenarios
- [ ] This was checked against keyboard-only and screenreader scenarios
- [ ] This required updates to Framer X components

cchaos commented:
Congrats on your first EUI PR! 😆 

As a quick etiquette rule for our PR's, the PR template that has the checklist needs to be kept in the PR summary. It helps remind of all the things that we need to add alongside what the PR is actually addressing like a changelog entry and adding/updating any tests. You can strike out any items that don't make sense though. But we do need keep the whole thing in the summary.

thompsongl commented:
Thanks! Could you update the comment regarding initial sorting behavior and client sort logic? The statements are now misleading, given your changes.

https://github.com/elastic/eui/blob/4e08b088a60e9b2d3cb3ab54328c754cdce2109a/src/components/basic_table/in_memory_table.js#L359-L362

lizozom commented:
@thompsongl done

cchaos commented on src/components/basic_table/in_memory_table.js:
```diff
@@ -191,11 +191,19 @@ export class EuiInMemoryTable extends Component {
+      && this.state.sortDirection === 'desc'
+      && sortDirection === 'asc') {
+      sortField = '';
+      sortDirection = '';
```
The problem I'm encountering is there's no way to turn this off. Meaning, if the table should *always* be sorted, there is no way to force it.

chandlerprall replied:
I feel most things in this category, we'd push for EuiBasicTable to be used directly. InMemoryTable intends to provide the most common, desired functionality.

That said, @lizozom did offer to put a configuration flag on the table to enable/disable this functionality. I don't have a preference between the two options and am happy to defer.

thompsongl replied:
I was wondering about this, also. It isn't clearly stated that the original behavior was meant to enforce a continuous sort state. Adding flag seems like a good option

cchaos replied:
+1 to adding a flag

lizozom replied:
So this behavior will be supported only for InMemoryTable, and if anyone wants this behavior in EuiBasicTable, they have to implement this on their own, right?

cchaos replied:
Well it should technically be **on by default** by InMemoryTable but yes any consumer building tables with the basic table or completely custom will have access to this prop too.

cchaos replied:
Well it should technically be **on by default** by InMemoryTable but yes any consumer building tables with the basic table or completely custom will have access to this prop too.

cchaos commented:
I'm not sure where the aria-labelling is hooked up, but it now needs to account for this new behavior, but only if that flag is used.

<img width="730" alt="screen shot 2019-02-25 at 11 47 25 am" src="https://user-images.githubusercontent.com/549577/53354159-3b534780-38f4-11e9-851c-b93c65e313aa.png">

thompsongl commented:
> I'm not sure where the aria-labelling is hooked up, but it now needs to account for this new behavior, but only if that flag is used.

This is all done in `table_header_cell.js`. I actually think the aforementioned flag will be required so it can be passed down into the composition components.

lizozom commented:
@cchaos @thompsongl how do you activate this flag?

Also, this behavior is for InMemoryTable, not for the basic Table component.

lizozom commented:
@cchaos @thompsongl 

I dug in a little deeper, and I think that having this work only for InMemTable and not for BasicTable creates odd behavior. It is also a problem since both implementations share the same table header cell implementation.

On the other hand, I'm not sure how to make BasicTable support this, since it can't restore the original order of items. Lets sync on this when you're available.

lizozom commented:
@cchaos  I updated the aria labels and added a flag so that BasicTable could support both modes (with and without neutral sort).

I didn't expose the flag in the InMemoryTable api yet. Should I? And if so, what should be the default?
Since we want InMemoryTable to showcase our common behavior, shouldn't we simplify it by choosing one?

cchaos commented on src/components/table/table_header_cell.js:
```diff
@@ -54,6 +55,16 @@ export const EuiTableHeaderCell = ({
+      } else if (allowNeutralSort && ariaSortValue === 'descending') {
+        return 'Click to unsort';
+      } else {
+        return 'Click to sort in ascending order';
```
🥇

cchaos commented on src/components/basic_table/in_memory_table.js:
```diff
@@ -352,12 +361,12 @@ export class EuiInMemoryTable extends Component {
         field: sortField,
         direction: sortDirection,
       },
+      allowNeutralSort: true,
```
I would set this to pass down the prop from line 75
```suggestion
      allowNeutralSort: allowNeutralSort,
```

and then set the prop to `true` as default after line 145 https://github.com/elastic/eui/blob/f02d7d804f0c80c15cda9fad97203b70845320b9/src/components/basic_table/in_memory_table.js#L145
`allowNeutralSort: true`

cchaos replied:
This way consumers can override the default by setting it to false.

cchaos commented on src/components/table/table_header_cell.js:
```diff
@@ -109,6 +120,7 @@ EuiTableHeaderCell.propTypes = {
   onSort: PropTypes.func,
   isSorted: PropTypes.bool,
   isSortAscending: PropTypes.bool,
+  allowNeutralSort: PropTypes.bool,
```
Can you add a comment above this line explaining what the prop does? This helps with our auto-documentation.

cchaos commented on src/components/basic_table/basic_table.js:
```diff
@@ -138,7 +138,8 @@ export const SelectionType = PropTypes.shape({
 const SortingType = PropTypes.shape({
-  sort: PropertySortType
+  sort: PropertySortType,
+  allowNeutralSort: PropTypes.bool,
```
Our props documentation for this component is in a slightly odd place. But could you add this option and description to this file/section:

https://github.com/elastic/eui/blob/e6e3168c7b988221388d480270ee058e5839a3a5/src-docs/src/views/tables/basic/props_info.js#L131-L142

cchaos commented on src/components/basic_table/in_memory_table.js:
```diff
@@ -72,6 +72,7 @@ const InMemoryTablePropTypes = {
       sort: PropertySortType
     })
   ]),
+  allowNeutralSort: PropTypes.bool,
```
Similar to the comment about the props documentation for basic table, there's one for the in memory table too: https://github.com/elastic/eui/blob/master/src-docs/src/views/tables/in_memory/props_info.js

lizozom commented:
@cchaos done review
Please check if I did the documentation right.

thompsongl commented on src/components/basic_table/in_memory_table.js:
```diff
@@ -191,11 +195,19 @@ export class EuiInMemoryTable extends Component {
     } = sort;
 
+    // Allow going back to 'neutral' sorting
+    if (this.state.sortField === sortField
```
`allowNeutralSort` needs to be part of this logic. Neutral sort still happens even if the flag is set to false

lizozom replied:
oops!
deleted this by mistake while testing something.

thompsongl commented on src/components/basic_table/in_memory_table.js:
```diff
@@ -352,12 +364,12 @@ export class EuiInMemoryTable extends Component {
         field: sortField,
         direction: sortDirection,
       },
+      allowNeutralSort: (this.props.allowNeutralSort === false) ? false : true,
```
Add `allowNeutralSort` to the destructured `this.props` object above. Right now it's included in `rest` and gets passed through to the rendered `div`, which gives a React warning.

`this.props.allowNeutralSort` can then change to just `allowNeutralSort`, also.

thompsongl replied:
Looking at this again, I see that you probably rely on `rest` to pass the flag to tableHeaderCell. If that's the case, we still need to exclude the flag from being passed to other parts of basicTable

lizozom replied:
I wasn't relying on the rest param, it's being passed in the sorting object.
Anyway, I had to include allowNeutralSorting in the state object, since props is not available in the onTableUpdate callback.

lizozom commented:
@cchaos @thompsongl we good to merge?

cchaos commented on CHANGELOG.md:
```diff
@@ -1,5 +1,6 @@
 ## [`master`](https://github.com/elastic/eui/tree/master)
 
+- Added `allowallowNeutralSort` prop to `EuiInMemoryTable` to support unsorting table columns ([#1591](https://github.com/elastic/eui/pull/1591))
```
Looks like a type in CL

lizozom commented:
Once it's in master, how can I start using it in Kibana?

chandlerprall commented:
@lizozom as soon as #1619 is merged we'll make an EUI release which will include these table changes. We'll create a PR in kibana to bring the latest EUI into master, and backport that into the 7.x branch.
